File: leet-code/46. Permutations/Permutations46.java
import java.util.ArrayList;
import java.util.List;

public class Permutations46 {
    public List<List<Integer>> permute(int[] nums) {
         return getPermutations(nums, new ArrayList<>(), new ArrayList<>());
    }

    private List<List<Integer>> getPermutations(int[] num, List<List<Integer>> perms, List<Integer> perm) {
        if (num.length == perm.size()) {
            perms.add(new ArrayList<>(perm));
            return perms;
        }

         for (int n : num) {
             if (!perm.contains(n)) {
                 perm.add(n);
                 getPermutations(num, perms, perm);
                 perm.remove(perm.size() - 1);
             }
         }
         return perms;
    }
}

File: leet-code/46. Permutations/Permutations46.kt
class Permutations46_2 {
    fun permute(nums: IntArray): List<List<Int>> {
        val perms: MutableList<List<Int>> = mutableListOf()
        val perm: MutableList<Int> = mutableListOf()

        fun dfs(): List<List<Int>> {
            if (perm.size == nums.size) {
                perms.add(ArrayList(perm))
                return perms.toList()
            }

            for (i in nums.indices) {
                if (perm.contains(nums[i])) continue

                perm.add(nums[i])
                dfs()
                perm.removeLast()
            }
            return perms.toList()
        }

        return dfs()
    }
}
